File: pset6/sentimental-credit/credit.py
from cs50 import get_string


def main():
    credit = get_string("Number: ")
    length = len(credit)
    temp = sum(length, credit)

    if length in [13, 16] and credit[0] in ["4"]:
        if temp % 10 == 0:
            print("VISA")
        else:
            print("INVALID")

    elif length in [15] and credit[0] in ["3"] and credit[1] in ["4", "7"]:
        if temp % 10 == 0:
            print("AMEX")
        else:
            print("INVALID")

    elif length in [16] and credit[0] in ["5"] and credit[1] in ["1", "2", "3", "4", "5"]:
        if temp % 10 == 0:
            print("MASTERCARD")
        else:
            print("INVALID")

    else:
        print("INVALID")


def sum(length, credit):
    temp = 0
    
    for n in range(length - 1, 0, -2):
        creditNum = int(credit[n - 1])
        creditStr = str(creditNum * 2)
        for l in range(len(creditStr)):
            temp = temp + int(creditStr[l])

    for n in range(length, 0, -2):
        creditNum = int(credit[n - 1])
        temp = temp + creditNum

    return temp


if __name__ == "__main__":
    main()

File: pset1/credit/credit.c
#include <cs50.h>
#include <stdio.h>
#include <string.h>

int ctoi(int digit, char credit[]);
int sum(int len, char credit[]);

int main(void)
{
    string creditString;
    creditString = get_string("Number: ");
    int len = strlen(creditString);

    char credit[len];
    strcpy(credit, creditString);

    int temp = sum(len, credit);

    if ((len == 13 || len == 16) && (credit[0] == '4'))
    {
        if (temp % 10 == 0)
        {
            printf("VISA\n");
        }
        else
        {
            printf("INVALID\n");
        }
    }
    else if ((len == 15) && (credit[0] == '3') && (credit[1] == '4' || credit[1] == '7'))
    {
        if (temp % 10 == 0)
        {
            printf("AMEX\n");
        }
        else
        {
            printf("INVALID\n");
        }
    }
    else if ((len == 16) && (credit[0] == '5') && (credit[1] == '1' || credit[1] == '2' || credit[1] == '3' || credit[1] == '4'
             || credit[1] == '5'))
    {
        if (temp % 10 == 0)
        {
            printf("MASTERCARD\n");
        }
        else
        {
            printf("INVALID\n");
        }
    }
    else
    {
        printf("INVALID\n");
    }
}

int ctoi(int digit, char credit[])
{
    int creditNum = (int)(credit[digit - 1]) - 48;
    return creditNum;
}

int sum(int len, char credit[])
{
    int temp = 0;
    for (int n = len - 1; n > 0; n -= 2)
    {
        int creditNum = ctoi(n, credit);
        temp = temp + (creditNum * 2) % 10 + (creditNum * 2) / 10;
        printf("%i ", temp);
    }
    for (int n = len; n > 0; n -= 2)
    {
        int creditNum = ctoi(n, credit);
        temp = temp + creditNum;
    }
    return temp;
}